marketo/errors.go

```diff
@@ -0,0 +1,154 @@
+	Message string `json:"message"`
+}
+
+type Respose struct {
+	Errors    []Error          `json:"errors"`
+	Result    []map[string]any `json:"result"`
+	RequestID string           `json:"requestId"`
+	Success   bool             `json:"success"`
+}
+
+// checkResponseLeverErr reports wheather the response level error is available or not.
+// If available, returns the error code as well.
+func checkResponseLeverErr(body []byte) (bool, int, error) {
+	var resp Respose
+	if err := json.Unmarshal(body, &resp); err != nil {
+		return false, 0, err
+	}
+
+	if len(resp.Errors) == 0 {
+		return false, 0, nil
+	}
+
+	code, err := strconv.Atoi(resp.Errors[0].Code)
+	if err != nil {
+		return false, 0, err
+	}
+
+	return (len(resp.Errors) > 0), code, nil
+}
+
+// InterpretError interprets the given HTTP response (in a fairly straightforward
+// way) and returns an error that can be handled by the caller.
+func interpretError(res *http.Response, body []byte) error { //nolint:cyclop
+	// A must check.
+	if res.StatusCode >= 200 && res.StatusCode <= 299 {
+		erroneous, code, err := checkResponseLeverErr(body)
+		if err != nil {
+			return err
+		}
+
+		// If response is 200 OK, but erroneous, we update the status code,
+		//  continue with the switch cases.
+		if erroneous {
+			statusCode := statusCodeMap(code)
+			res.StatusCode = statusCode
+		} else {
+			return nil
+		}
+	}
+
+	switch res.StatusCode {
+	case http.StatusUnauthorized:
+		// Access token invalid, refresh token and retry
+		return common.NewHTTPStatusError(res.StatusCode, fmt.Errorf("%w: %s", common.ErrAccessToken, string(body)))
+	case http.StatusForbidden:
+		// Forbidden, not retryable
+		return common.NewHTTPStatusError(res.StatusCode, fmt.Errorf("%w: %s", common.ErrForbidden, string(body)))
+	case http.StatusNotFound:
+		// Semantics are debatable (temporarily missing vs. permanently gone), but for now treat this as a retryable error
+		return common.NewHTTPStatusError(res.StatusCode, fmt.Errorf("%w: %s", common.ErrRetryable, string(body)))
+	case http.StatusTooManyRequests:
+		// Too many requests, retryable
+		return common.NewHTTPStatusError(res.StatusCode, fmt.Errorf("%w: %s", common.ErrRetryable, string(body)))
+	}
+
+	if res.StatusCode >= 400 && res.StatusCode < 500 {
+		return common.NewHTTPStatusError(res.StatusCode, fmt.Errorf("%w: %s", common.ErrCaller, string(body)))
+	} else if res.StatusCode >= 500 && res.StatusCode < 600 {
+		return common.NewHTTPStatusError(res.StatusCode, fmt.Errorf("%w: %s", common.ErrServer, string(body)))
+	}
+
+	return common.NewHTTPStatusError(res.StatusCode, fmt.Errorf("%w: %s", common.ErrUnknown, string(body)))
+}
+
+// statusCodeMap maps the erroneous response from marketo, with a valid http status code.
+// The response body can be sent as is.
+// https://experienceleague.adobe.com/en/docs/marketo-developer/marketo/rest/error-codes
+func statusCodeMap(code int) int { //nolint:funlen,cyclop
+	switch code {
+	case 502: //nolint:gomnd
```

Comment:
This must have been painful, thanks for the hard work! :) A quick tip would be to use a function level no lint (just add it one line before the function) to disable the check for the entire function, instead of adding it per line. 

ref - https://github.com/golangci/golangci-lint/discussions/1895